test/ExecuteUpdateTest.cpp

```diff
@@ -32,6 +32,165 @@ MATCHER_P(AlwaysFalse, msg, "") {
+      localVocabEntries;
+  auto LVI = [&localVocabEntries](const std::string& iri) {
+    if (!localVocabEntries.contains(iri)) {
+      localVocabEntries.try_emplace(
+          iri, std::make_unique<LocalVocabEntry>(
+                   ad_utility::triple_component::Iri::fromIriref(iri)));
+    }
+    return Id::makeFromLocalVocabIndex(localVocabEntries.at(iri).get());
+  };
+
+  auto IdTriple = [defaultGraphId](const ::Id s, const ::Id p, const ::Id o,
+                                   const std::optional<::Id> graph =
+                                       std::nullopt) -> ::IdTriple<> {
+    return ::IdTriple({s, p, o, graph.value_or(defaultGraphId)});
+  };
+
+  auto executeComputeGraphUpdateQuads = [&qec,
+                                         &index](const std::string& update) {
+    const auto sharedHandle =
+        std::make_shared<ad_utility::CancellationHandle<>>();
+    const std::vector<DatasetClause> datasets = {};
+    auto pq = SparqlParser::parseQuery(update);
+    QueryPlanner qp{qec, sharedHandle};
+    const auto qet = qp.createExecutionTree(pq);
+    return ExecuteUpdate::computeGraphUpdateQuads(index, pq, qet, sharedHandle);
+  };
+  auto expectComputeGraphUpdateQuads =
+      [&executeComputeGraphUpdateQuads](
+          const std::string& update,
+          const testing::Matcher<const std::vector<::IdTriple<>>&>&
+              toInsertMatcher,
+          const testing::Matcher<const std::vector<::IdTriple<>>&>&
+              toDeleteMatcher) {
+        EXPECT_THAT(
+            executeComputeGraphUpdateQuads(update),
+            testing::Pair(AD_FIELD(ExecuteUpdate::IdTriplesAndLocalVocab,
+                                   idTriples, toInsertMatcher),
+                          AD_FIELD(ExecuteUpdate::IdTriplesAndLocalVocab,
+                                   idTriples, toDeleteMatcher)));
+      };
+  auto expectComputeGraphUpdateQuadsFails =
+      [&executeComputeGraphUpdateQuads](
+          const std::string& update,
+          const testing::Matcher<const std::string&>& messageMatcher) {
+        AD_EXPECT_THROW_WITH_MESSAGE(executeComputeGraphUpdateQuads(update),
+                                     messageMatcher);
+      };
+
+  expectComputeGraphUpdateQuads(
+      "INSERT DATA { <s> <p> <o> . }",
+      testing::ElementsAreArray({IdTriple(LVI("<s>"), LVI("<p>"), LVI("<o>"))}),
+      testing::IsEmpty());
+  expectComputeGraphUpdateQuads(
+      "DELETE DATA { <z> <label> \"zz\"@en }", testing::IsEmpty(),
+      testing::ElementsAreArray(
+          {IdTriple(Id("<z>"), Id("<label>"), Id("\"zz\"@en"))}));
+  expectComputeGraphUpdateQuads(
+      "DELETE { ?s <is-a> ?o } INSERT { <s> <p> <o> } WHERE { ?s <is-a> ?o }",
+      testing::ElementsAreArray({IdTriple(LVI("<s>"), LVI("<p>"), LVI("<o>")),
+                                 // TODO: what to do with this duplicate entry?
+                                 IdTriple(LVI("<s>"), LVI("<p>"), LVI("<o>"))}),
```

Comment:
I don't understand the question?

Reply:
That was a note that I left in. It was meant to refer to the fact that Quads may generated multiple times by `computeGraphUpdateQuads`. This may happen because an interpolated variable has the same value in multiple rows or with quads that have no variables at all. They are currently dropped during processing by DeltaTriples. One might want to optimise some of the cases earlier. But that can happen later on.